### DISCUSSION TITLE
How to determine if `ConstFetch` is used as default value of property/param

### DISCUSSION THREAD

**Discussion by Wirone** [rectorphp/rector#8208] [Q&A] [answered]
I have custom rector rule with such logic:

```php
private function shouldSkip(ConstFetch $node): bool
{
    if (!in_array($node->name->toString(), $this->constNames, true)) {
        return true;
    }

    // Ignore if constant is used as default value for class properties' or method param's default value
    // because expressions are not allowed there as of now.
    $parent = $node->getAttribute('parent');

    return !is_object($parent) || in_array($parent::class, [Param::class, PropertyProperty::class]);
}
```

It works on 0.18.0, but fails when I upgrade to 0.18.3 - I believe it's related to the process of removing parent node concept. But in this case it was the best approach to verify the context where the const is fetched... Should I change the rector's `getNodeTypes()` to listen to `Param` and `PropertyProperty` and check it from the other side, or is there better way to do it?

**Comment by samsonasik** [comment, marked as answer]
Hi, you can add `Param` and `PropertyProperty` to `getNodeTypes()` sorted before `ConstFetch` definiton:

```php
public function getNodeTypes(): array
{
    return [Param::class, PropertyProperty::class, ConstFetch::class];
}
```

Then check on `refactor()` method early:

```php
public function refactor(Node $node) {
    if ($node instanceof Param || $node instanceof  PropertyProperty) {
        return \PhpParser\NodeTraverser::DONT_TRAVERSE_CURRENT_AND_CHILDREN;
    } 

    // else process the ConstFetch that not in param or propertyproperty
```

Above is pseudocode, but you got the idea :)

**Reply by Wirone** [reply]
It works, thank you! I did not realise `refactor()` changed its return type, I had `?Node` in my custom rule and was confused by `return \PhpParser\NodeTraverser::DONT_TRAVERSE_CURRENT_AND_CHILDREN;`, but then I checked it out and yeah, confirmed 😁.
